### DISCUSSION TITLE
How does one handle SIGTERM?

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#162] [Q&A]
The POSIX signal `SIGTERM` is the default `kill(1)` action to terminate a program cleanly; essentially `BYE`.  I do not see a suitable exception code for this.  Should the system define one of its own from the reserved system exceptions [-4095..-256]?  Or should there be a standard one assigned?  Also its behaviour when there is no `CATCH` frame would be to cleanly exit in the same manner as `BYE`.

**Comment by ruv** [comment]
> The POSIX signal SIGTERM is the default kill(1) action to terminate a program cleanly;

There is no a standard throw code for this case at the moment, so the system should define own code from the range [-4095..-256].

A standard throw code could be assigned in the future.

>  Also its behaviour when there is no `CATCH` frame would be to cleanly exit in the same manner as `BYE`.

Yes. But the [_exit status_](https://en.wikipedia.org/wiki/Exit_status) should not be 0 in this case. Probably, it makes sense to always use the throw code as the _exit status_ by default (i.e., unless the program specifies a different value for the _exit status_ using system-specific means).

A Forth system can support a batch mode, which avoids any interaction with the user. In this mode, REPL (interpreting of _stdin_) should not show a prompt, should output error messages in _stderr_, and should stop on any uncaught exception.  It's convenient to use the throw code as the _exit status_ by default in this mode.

**Reply by SirWumpus** [reply]
> Probably, it makes sense to always use the throw code as the exit status by default

Not practical since the POSIX exit code only uses the 8 least significant bits.  Plus the shell will use codes 128+.  From NetBSD `man exit`:
```
RESTRICTIONS
     Standard C guarantees only that the values zero, EXIT_SUCCESS, and
     EXIT_FAILURE produce meaningful results.  POSIX extends this to guarantee
     that the least significant 8 bits of status are preserved and returned to
     the parent via wait(2).  Values outside the supported range 0-255 are
     bitwise-truncated; therefore, negative values should not be used.
```
So in light of that I'd probably opt for `1 bye-code` in order that pipe is terminated.

**Reply by SirWumpus** [reply]
I recall the shell using `signal number + 128`.

**Reply by ruv** [reply]
Ah, I see. Then, system-defined (or program-defined, if any) set of codes should be used.

**Reply by ruv** [reply]
> I'd probably opt for `1 bye-code` in order that pipe is terminated.

"code" have a lot of meanings. Maybe `bye-status` is better.  SP-Forth uses `halt ( n -- ⊥ )`

**Reply by SirWumpus** [reply]
I had considered `HALT` as a name, but felt it might be confused with other uses of halt, ie. `halt(1)`, `HLT` machine instruction, or [HCF](https://en.wikipedia.org/wiki/Halt_and_Catch_Fire_(computing)).  I would have preferred `EXIT` but that has already been standardised.  So I choose to remain close to the `BYE` name.  `BYE-STATUS` have to think on (exit code vs exit status).
